File: Project 1/target.py
"""
target.py
=========
Defines the data structures used to store and represent
assembly instructions after code generation.

Pipeline:
===============
(Sits at the end of the pipeline as the final output stage:)

  (1)  interference.py  <- provides register assignments.
          
  (2)  target.py        <- stores and formats the generated assembly instructions.
          
  (3)  output           <- assembly code ready to be written to a .s file.

Responsibilities:
===========================
- Define AsmInstruction, representing a single assembly instruction.
- Define TargetCode, representing a full sequence of assembly instructions.
- Provide string output that formats instructions in correct assembly syntax.

Associated Dependencies:
==========================
NA

Usage Example:
===================
NA

Misc Notes:
============
NA
"""

from typing import List


class AsmInstruction:
    """
    AsmInstruction: 

    Represents a single assembly language instruction.

    An instruction consists of:
        - an opcode (e.g., ADD, SUB, MUL, DIV, MOV)
        - an optional source operand
        - an optional destination operand

    Supported forms:
        opcode
        opcode src
        opcode src,dst

    Examples:
        ADD #1,R0
        MOV R1,a
        MUL R2,R3
    """
    def __init__(self, opcode: str, src: str = None, dst: str = None, UnaryNeg: bool = False):
        self.opcode = opcode
        self.src = src if not UnaryNeg else -src #might be an incorrect implementation. Double check.
        self.dst = dst

    def __str__(self) -> str:
        if self.src is None and self.dst is None:
            return self.opcode
        if self.dst is None:
            return f"{self.opcode} {self.src}"
        return f"{self.opcode} {self.src},{self.dst}"



class TargetCode:
    """
    TargetCode: 

    Represents a sequence of assembly instructions corresponding to a single basic block of generated target code.

    This class serves as a container for AsmInstruction objects and provides support routines for constructing and printing the
    target code sequence.
    """
    def __init__(self) -> None:
        self.instructions: List[AsmInstruction] = []

    def add(self, instr: AsmInstruction) -> None:
        self.instructions.append(instr)             # Append an assembly instruction to the target code sequence.

    def __str__(self) -> str:
        """
        Return the complete target code as a newline-separated string
        suitable for writing to an output .s file.
        """
        return "\n".join(str(i) for i in self.instructions)


        

File: COMP3649-project-1--main/Haskell-Solution/src/Target.hs
{-
  Target.hs — Maps to: target.py

  Defines data structures for storing and displaying
  target assembly instructions after code generation.

  Role in the Pipeline
  --------------------
  Sits at the end of the pipeline as the final output stage:

    Interference.hs  <- provides register assignments
          |
    Target.hs        <- stores and formats the generated assembly instructions
          |
    output           <- assembly code ready to be written to a .s file

  Responsibilities
  ----------------
  - Define AsmInstruction, representing a single assembly instruction.
  - Define TargetCode, representing a full sequence of assembly instructions.
  - Provide string output that formats instructions in correct assembly syntax.
  - Hide data constructors; provide a clean smart-constructor interface.

  Out of Scope
  ------------
  - Parsing input files (Parser.hs).
  - Computing liveness (Liveness.hs).
  - Building interference graphs or assigning registers (Interference.hs).
  - Translating three-address instructions (CodeGen.hs).

  Key Abstractions
  ----------------
  AsmInstruction
      One assembly instruction: opcode, optional src, optional dst.

  TargetCode
      Full sequence of assembly instructions for a basic block.
-}

module Target
  ( AsmInstruction
  , TargetCode
  , mkAsmInstruction
  , getOpcode, getSrc, getDst
  , emptyTargetCode, addInstruction, addInstructions
  , getInstructions
  , showAsmInstruction, showTargetCode
  ) where


-- ============================================================
-- AsmInstruction: one assembly instruction
-- ============================================================

{-
  Internal representation:
    Asm opcode (Maybe src) (Maybe dst)

  Supported forms:
    opcode              e.g. NOP
    opcode src          e.g. PUSH R0
    opcode src,dst      e.g. ADD #1,R0   MOV R1,a
-}

data AsmInstruction = Asm String (Maybe String) (Maybe String)
  deriving (Show)


-- | Constructor for all instruction forms.
mkAsmInstruction :: String -> Maybe String -> Maybe String -> AsmInstruction
mkAsmInstruction = Asm


-- ------------------------------------------------------------
-- Accessors
-- ------------------------------------------------------------

getOpcode :: AsmInstruction -> String
getOpcode (Asm o _ _) = o

getSrc :: AsmInstruction -> Maybe String
getSrc (Asm _ s _) = s

getDst :: AsmInstruction -> Maybe String
getDst (Asm _ _ d) = d


-- ------------------------------------------------------------
-- Display
-- ------------------------------------------------------------

showAsmInstruction :: AsmInstruction -> String
showAsmInstruction (Asm op Nothing  Nothing)  = op
showAsmInstruction (Asm op (Just s) Nothing)  = op ++ " " ++ s
showAsmInstruction (Asm op (Just s) (Just d)) = op ++ " " ++ s ++ "," ++ d
showAsmInstruction (Asm op Nothing  (Just _)) = op  -- dst without src: show opcode only


-- ============================================================
-- TargetCode: full sequence of assembly instructions
-- ============================================================

data TargetCode = TCode [AsmInstruction]
  deriving (Show)


-- | Create an empty target code sequence.
emptyTargetCode :: TargetCode
emptyTargetCode = TCode []

-- | Append a single instruction to the end of the sequence.
addInstruction :: AsmInstruction -> TargetCode -> TargetCode
addInstruction i (TCode is) = TCode (is ++ [i])

-- | Append a list of instructions to the end of the sequence.
addInstructions :: [AsmInstruction] -> TargetCode -> TargetCode
addInstructions new (TCode is) = TCode (is ++ new)

-- | Return the list of instructions.
getInstructions :: TargetCode -> [AsmInstruction]
getInstructions (TCode is) = is

-- | Format the full target code for output to a .s file.
showTargetCode :: TargetCode -> String
showTargetCode (TCode is) = unlines (map showAsmInstruction is)


{-
  GHCi test commands:

  :load Target
  showAsmInstruction (mkAsmInstruction "ADD" (Just "#1") (Just "R0"))
  -- "ADD #1,R0"

  let t = addInstructions
            [ mkAsmInstruction "MOV" (Just "a") (Just "R0")
            , mkAsmInstruction "ADD" (Just "#1") (Just "R0")
            ] emptyTargetCode
  putStr (showTargetCode t)
  -- MOV a,R0
  -- ADD #1,R0
-}
